src/actions/clipboard.ts

```diff
@@ -335,54 +360,62 @@ export class Clipboard {
+    let workspace;
+    if (focusTree instanceof WorkspaceSvg) {
+      workspace = focusTree;
+    } else if (focusTree instanceof Flyout) {
+      // Seems like this case doesn't actually happen and a
+      // (flyout) Workspace is returned instead, but it's possible
+      workspace = focusTree.targetWorkspace;
+    } else {
+      // Give up and just paste in the main workspace
+      workspace = getMainWorkspace() as WorkspaceSvg;
+    }
 
-    return this.navigation.canCurrentlyEdit(workspace) && !Gesture.inProgress();
+    if (!workspace) return null;
+    // If we're trying to paste in a flyout, paste in the target workspace instead
+    if (workspace.isFlyout)
+      workspace = workspace.targetWorkspace as WorkspaceSvg;
+
+    return workspace;
   }
 
   /**
-   * Callback function for pasting a block from keyboard
-   * navigation. This callback is shared between keyboard shortcuts
-   * and context menu items.
+   * The callback for the paste action. Uses the registered version of the paste callback
+   * to perform the paste logic, then clears any toasts about pasting.
    *
-   * @param workspace The `WorkspaceSvg` where the shortcut was
-   *     invoked.
-   * @returns True if this function successfully handled pasting.
+   * @param workspace Workspace where shortcut happened.
+   * @param e menu open event or keyboard event
+   * @param shortcut keyboard shortcut or undefined for context menus
+   * @param scope scope of the shortcut or context menu item
+   * @returns true if a paste happened, false otherwise
    */
-  private pasteCallback(workspace: WorkspaceSvg) {
-    if (!this.copyData || !this.copyWorkspace) return false;
+  private pasteCallback(
+    workspace: WorkspaceSvg,
+    e: Event,
+    shortcut: ShortcutRegistry.KeyboardShortcut = {
+      name: Constants.SHORTCUT_NAMES.CUT,
+    },
+    scope: ContextMenuRegistry.Scope,
+  ) {
+    const didPaste =
+      !!this.oldPasteCallback &&
+      this.oldPasteCallback(workspace, e, shortcut, scope);
+
+    // Clear the paste hints regardless of whether something was pasted
+    // Some implementations of paste are async and we should clear the hint
+    // once the user initiates the paste action.
     clearPasteHints(workspace);
-
-    const pasteWorkspace = this.copyWorkspace.isFlyout
-      ? workspace
-      : this.copyWorkspace;
-
-    const targetNode = FocusableTreeTraverser.findFocusedNode(
-      pasteWorkspace,
-    ) as unknown as INavigable<any>;
-    // If we're pasting in the flyout it still targets the workspace. Focus
-    // first as to ensure correct selection handling.
-    getFocusManager().focusTree(workspace);
-
-    Events.setGroup(true);
-    const block = clipboard.paste(this.copyData, pasteWorkspace) as BlockSvg;
-    if (block) {
-      if (targetNode) {
-        this.navigation.tryToConnectBlock(targetNode, block);
-      }
-      Events.setGroup(false);
-      return true;
-    }
-    Events.setGroup(false);
```

Comment:
This removed part doesn't have an equivalent so the behaviour has reverted to the core behaviour that doesn't connect.

I'm not sure this is a bad thing - see discussion on #480 and recent user testing feedback - but checking whether it's intended here.

Reply:
Yes, I read #480 and figured that staying consistent with the behavior in core made sense, and if we wanted to change it, we could change it in core. It should be easy to do a one-time move mode toast, I should get to the "keyboard mode" heuristic soon and we could use that.